controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,588 @@
+	for m.processNextWorkItem(id) {
+	}
+}
+
+func (m *SnapshotMonitor) Close() {
+	m.logger.WithField("monitor", monitorName).Info("Closing snapshot monitor")
+
+	m.snapshotCheckTaskQueue.ShutDown()
+	m.checkScheduler.Stop()
+	m.quit()
+}
+
+func (m *SnapshotMonitor) Update(map[string]interface{}) error {
+	dataIntegrityCronJob, err := m.ds.GetSettingValueExisted(types.SettingNameSnapshotDataIntegrityCronJob)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get %v setting", types.SettingNameSnapshotDataIntegrityCronJob)
+	}
+
+	if dataIntegrityCronJob == m.existingDataIntegrityCronJob && m.checkScheduler.Len() > 0 {
+		return nil
+	}
+
+	if m.checkScheduler.Len() > 0 {
+		m.checkScheduler.Remove(m.checkSnapshots)
+	}
+
+	job, err := m.checkScheduler.Cron(dataIntegrityCronJob).Do(m.checkSnapshots)
+	if err != nil {
+		return errors.Wrap(err, "failed to schedule snapshot check job")
+	}
+
+	m.Lock()
+	m.existingDataIntegrityCronJob = dataIntegrityCronJob
+	m.Unlock()
+
+	m.checkScheduler.StartAsync()
+
+	m.logger.WithField("monitor", monitorName).Infof("Next snapshot check job will be executed at %v", job.NextRun())
+
+	return nil
+}
+
+func (m *SnapshotMonitor) GetData() (interface{}, error) {
+	m.RLock()
+	defer m.RUnlock()
+	return m.SnapshotMonitorStatus, nil
+}
+
+func (m *SnapshotMonitor) addToInProgressSnapshotCheckTasks(snapshotName string) bool {
+	m.inProgressSnapshotCheckTasksLock.Lock()
+	defer m.inProgressSnapshotCheckTasksLock.Unlock()
+
+	_, ok := m.inProgressSnapshotCheckTasks[snapshotName]
+	if ok {
+		m.logger.WithField("monitor", monitorName).Debugf("snapshot %s is being checked", snapshotName)
+		return false
+	}
+	m.inProgressSnapshotCheckTasks[snapshotName] = struct{}{}
+
+	return true
+}
+
+func (m *SnapshotMonitor) deleteFromInProgressSnapshotCheckTasks(snapshotName string) {
+	m.inProgressSnapshotCheckTasksLock.Lock()
+	defer m.inProgressSnapshotCheckTasksLock.Unlock()
+
+	delete(m.inProgressSnapshotCheckTasks, snapshotName)
+}
+
+func (m *SnapshotMonitor) RunTask(arg interface{}) error {
+	task, ok := arg.(snapshotCheckTask)
+	if !ok {
+		return fmt.Errorf("failed to assert value: %v", arg)
+	}
+
+	if !m.addToInProgressSnapshotCheckTasks(task.snapshotName) {
+		return nil
+	}
+	defer m.deleteFromInProgressSnapshotCheckTasks(task.snapshotName)
+
+	engine, err := m.ds.GetVolumeCurrentEngine(task.volumeName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get engine for volume %v", task.volumeName)
+	}
+
+	if err := m.checkVolumeNotInMigration(engine.Spec.VolumeName); err != nil {
+		return err
+	}
+
+	engineCliClient, err := m.getEngineBinaryClient(engine.Spec.VolumeName)
+	if err != nil {
+		return err
+	}
+
+	engineClientProxy, err := engineapi.GetCompatibleClient(engine, engineCliClient, m.ds, m.logger, m.proxyConnCounter)
+	if err != nil {
+		return err
+	}
+	defer engineClientProxy.Close()
+
+	m.Lock()
+	m.SnapshotCheckState = longhorn.SnapshotCheckStateInProgress
```

Comment:
I don't get why the monitor needs this state

Reply:
I'm thinking if we need to show `The node is doing snapshot checksum calculation` on node ui page, and the users can expect the system might slow down.
If no, we don't need the state.
WDYT?

Reply:
It's nice to have. But the concern is, the monitor on this node issuing hash requests does not mean the actual calculation is on this node. 
BTW, why not use a counter instead? There can be 5 workers executing hash tasks simultaneously. Directly updating this field may be an issue.

Reply:
Yes, the concern is fair. I will probably remove it, bacause it will make users confused.